rust/crates/langsmith-pyo3/src/py_run.rs

```diff
@@ -46,43 +46,40 @@ impl FromPyObject<'_> for RunCreateExtended {
     }
 }
 
-fn extract_attachments(value: &Bound<'_, PyAny>) -> PyResult<Option<Vec<Attachment>>> {
-    if value.is_none() {
-        return Ok(None);
-    }
-
-    let mapping = value.downcast::<PyMapping>()?;
+#[derive(Debug)]
+pub struct RunUpdateExtended(langsmith_tracing_client::client::RunUpdateExtended);
 
-    let size = mapping.len()?;
-    if size == 0 {
-        return Ok(None);
+impl RunUpdateExtended {
+    #[inline]
+    pub(crate) fn into_inner(self) -> langsmith_tracing_client::client::RunUpdateExtended {
+        self.0
     }
+}
 
-    let mut attachments = Vec::with_capacity(size);
-
-    for result in mapping.items()?.iter()? {
-        let key_value_pair = result?;
+impl FromPyObject<'_> for RunUpdateExtended {
+    fn extract_bound(value: &Bound<'_, PyAny>) -> PyResult<Self> {
+        let run_update = value.extract::<RunUpdate>()?.into_inner();
 
-        let key_item = key_value_pair.get_item(0)?;
-        let key = key_item.extract::<&str>()?;
+        let attachments = {
+            if let Ok(attachments_value) = value.get_item(pyo3::intern!(value.py(), "attachments"))
+            {
+                extract_attachments(&attachments_value)?
+            } else {
+                None
+            }
+        };
 
-        // Each value in the attachments dict is a (mime_type, bytes) tuple.
-        let value = key_value_pair.get_item(1)?;
-        let value_tuple = value.downcast_exact::<PyTuple>()?;
-        let mime_type_value = value_tuple.get_item(0)?;
-        let bytes_value = value_tuple.get_item(1)?;
+        let io = RunIO {
+            inputs: serialize_optional_dict_value(value, pyo3::intern!(value.py(), "inputs"))?,
+            outputs: serialize_optional_dict_value(value, pyo3::intern!(value.py(), "outputs"))?,
+        };
 
-        attachments.push(Attachment {
-            // TODO: It's unclear whether the key in the attachments dict is
-            //       the `filename`` or the `ref_name`, and where the other one is coming from.
-            ref_name: key.to_string(),
-            filename: key.to_string(),
-            data: bytes_value.extract()?,
-            content_type: mime_type_value.extract()?,
-        });
+        Ok(Self(langsmith_tracing_client::client::RunUpdateExtended {
+            run_update,
+            io,
+            attachments,
+        }))
     }
-
-    Ok(Some(attachments))
 }
```

Comment:
I just moved this function down, to keep all the type definitions next to each other in the file. Otherwise it remained unchanged.